### PR TITLE
Checkoutv2: Refactor sidebar promotions into CheckoutSidebarNudge

### PR DESCRIPTION
**Pull request by jjchrisdiehl**
This PR consolidates a number of separate sidebar promotional components into one 'CheckoutSidebarNudge' wrapper component. There are a number of components that should be refactored, but this PR specifically handles plan interval upsells - see here for full list https://github.com/Automattic/payments-shilling/issues/2422#issuecomment-1965647926

Part of these changes include changing the upsell component from a CTA button to a word bubble with an inline text link CTA:

| v1 | v2 |
| ----- | ----- |
| image | image |

| v1 | v2 |
| ----- | ----- |
| image | image |

Related to https://github.com/Automattic/payments-shilling/issues/2422

## Proposed Changes

* Update the `PromoCard` component to a newer `CheckoutPromoCard` which can handle multiple upgrade levels
* Move the monthly -> annual upsell from the Features List to the new checkout word bubble
* Update the `onUpgradeClick` function to handle more than annual -> biennial upgrades
* Styling changes to the word bubble in v2 design along with handling style regressions of other sidebar components such as the ones found in `SecondaryCartPromotions`
* Change `CheckoutSidebarNudgeWrapper` to flex and remove margin from nudge components

## Testing Instructions

Since the onUpgradeClick function has been modified for both checkout v1 (currently in production) and v2, we'll need to check that the function works as expected in both checkout flows - so the following tests need to be done for both v1 and v2 of checkout

**Test DotCom plans**
* Add a monthly plan to your cart and go to checkout - you should see an upsell nudge that reads `Longer plan billing cycles save you money and include a custom domain for free for the first year.`
* Add a monthly plan and a domain, you should be prompted to switch to an annual plan to receive your domain free for the first year (including your domain being listed in the sidebar) `Save money with longer billing cycles and get %(domainMeta)s free for the first year by {{upgradeLink}}switching to an annual plan.{{/upgradeLink}}`
* Click on sidebar upgrade link to switch to annual plan
* Now that you have an annual plan in the cart you should see the following nudge `Save up to %(savingsPercentage)s% on longer billing cycles! It's hassle-free and easy on your wallet. {{upgradeLink}}Switch to a two-year plan and save.{{/upgradeLink}}`, click on sidebar upgrade link to switch to two year plan
* At this stage we stop showing upgrade nudges for biennial -> triennial plans terms.

**Test Jetpack and Akismet Plans**
* Jetpack and Akismet do not use the new word bubble upsell yet, so at this time please add a Jetpack or Akismet product and ensure the old nudge is still showing as expected

**DIFM**
* Go to `http://calypso.localhost:3000/start/do-it-for-me/new-or-existing-site` and continue through the onboarding steps until you reach checkout
* You should see in the sidebar the DIFM `What's next?` stepper and no other upsell nudges

**Upcoming Renewals**
* This one is trickier, you'll need a site with multiple purchases from around the same date with similar expirations
* When you go to manually renew one of your purchases, Checkout should prompt you renew the others

### CODE REVIEW

File: client/my-sites/checkout/src/components/checkout-sidebar-plan-upsell/index.tsx
Comment (jjchrisdiehl) [inline client/my-sites/checkout/src/components/checkout-sidebar-plan-upsell/index.tsx]: This one is a bit tricky to figure out because I haven't found a way to get to this point without a plan in the cart, since code in CheckoutSidebarPlanUpsell relies on a plan existing in the cart.

File: client/my-sites/checkout/src/components/checkout-sidebar-plan-upsell/index.tsx
Comment (jjchrisdiehl) [inline client/my-sites/checkout/src/components/checkout-sidebar-plan-upsell/index.tsx]: Not sure if this is really needed, but I included it to ensure we return without showing the promo if we have no label to show. I could probably just lose the condition and cover all non monthly/annual use cases.

File: client/my-sites/checkout/src/components/test/secondary-cart-promotions.tsx
Comment (jjchrisdiehl) [inline client/my-sites/checkout/src/components/test/secondary-cart-promotions.tsx:44]: I believe this is correct, since I moved isPurchaseRenewal outside of the component which caused the test to be lacking this data (since I never added the new prop to the test)

**Comment by jjchrisdiehl** [comment]
One thing I couldn't quite figure out was how to get the call to action links to wrap between words. I'm not sure if this is a quirk of an anchor link within flexbox?

File: client/my-sites/checkout/src/components/checkout-main-content.tsx
Comment (sirbrillig) [inline client/my-sites/checkout/src/components/checkout-main-content.tsx]: Not a blocker but you don't need to memoize this; calculating it is not expensive and it's a boolean so using it as a reference anywhere is stable even across re-renders (true and false are always unchanged unless they actually change).

File: client/my-sites/checkout/src/components/checkout-sidebar-plan-upsell/index.tsx
Comment (sirbrillig) [reply client/my-sites/checkout/src/components/checkout-sidebar-plan-upsell/index.tsx]: I'm trying to grok why this is here. Before this PR, the logic was to show this notice if `plan && shouldUseCheckoutV2`, but now we show it if `! plan && domainRegistrationOrTransferInCart`? I know that this component isn't rendered unless `shouldUseCheckoutV2` is true, but why invert the `plan` condition and add `domainRegistrationOrTransferInCart`?

File: client/my-sites/checkout/src/components/checkout-sidebar-plan-upsell/index.tsx
Comment (sirbrillig) [inline client/my-sites/checkout/src/components/checkout-sidebar-plan-upsell/index.tsx]: If `labelText` is undefined, maybe we should not render the promo card at all? From the above conditions it looks like that would happen if something was not monthly, yearly, biennial, or triennial (maybe a one-time purchase or a purchase with a four-year renewal cycle). 

I guess that's what you are saying in your comment above. I agree, there's no need to check for biennial or triennial. Let's just check for `labelText`.

File: client/my-sites/checkout/src/components/checkout-sidebar-plan-upsell/index.tsx
Comment (sirbrillig) [inline client/my-sites/checkout/src/components/checkout-sidebar-plan-upsell/index.tsx:204]: Not a blocker but using `let` is risky for a couple reasons:

1. It allows future developers to continue to redefine this variable anywhere below this line, making it hard to tell what its value will be when we see an instance of it in the code; we have to read all the code between that use-case and this definition.
2. It means several different definitions might take place, meaning that the order of those definitions is critical to understand when reading the code. You have to always consider what might come below the current line any time there's a redefinition of the variable.

Ternaries are a good pattern to use to allow using `const` for a variable that can have two possible values, but here you have three possible values and ternaries are also not great for long values or complex conditions. In this case, you could use an IIFE instead:

```js
const newPlan = (() => {
		if ( currentVariant.termIntervalInMonths === 1 ) {
			return {
				product_slug: annualVariant.productSlug,
				product_id: annualVariant.productId,
			};
		}

		if ( currentVariant.termIntervalInMonths === 12 ) {
			return {
				product_slug: biennialVariant.productSlug,
				product_id: biennialVariant.productId,
			};
		}
})()

if ( ! newPlan ) {
	return;
}
```

**Review by sirbrillig — COMMENTED** [review]
I think this looks pretty good to me! I have one question and a couple of suggestions.

Also, your before/after screenshots are I think showing v1 checkout in the "Before" column and v2 checkout in the "After" column, which is kind of confusing. Maybe you can include a before/after for v1 and a before/after for v2 so it's easy to see what's changed?

File: client/my-sites/checkout/src/components/checkout-sidebar-plan-upsell/index.tsx
Comment (jjchrisdiehl) [reply client/my-sites/checkout/src/components/checkout-sidebar-plan-upsell/index.tsx]: Ah, so this is an attempt to refactor the 'free plan upsell' and move it from SecondaryCartPromotions into this section, since its effectively just a billing term upsell/upgrade.

I think shimming it into this new component would require a larger rewrite since all of the current conditions require a plan and this one requires the absence of a plan. Unfortunately, trying to support v1 `PromoCard`s functionality requires us to do the opposite of what I am trying to do here and it creates a logic nightmare. I may hold off on moving the free plan upsell until after v2 is merged and theres less competing logic.

File: client/my-sites/checkout/src/components/checkout-sidebar-plan-upsell/index.tsx
Comment (jjchrisdiehl) [reply client/my-sites/checkout/src/components/checkout-sidebar-plan-upsell/index.tsx]: Removed for now with a comment to add it back in after v2 merger.

File: client/my-sites/checkout/src/components/checkout-sidebar-plan-upsell/index.tsx
Comment (jjchrisdiehl) [reply client/my-sites/checkout/src/components/checkout-sidebar-plan-upsell/index.tsx:204]: Updated to use the IIFE, thanks for the example!

File: client/my-sites/checkout/src/components/checkout-main-content.tsx
Comment (jjchrisdiehl) [reply client/my-sites/checkout/src/components/checkout-main-content.tsx]: Ah good point, I just moved it from another section where it was memoized, but I suppose it was probably being recalculated more often in its original placement?

Comment (sirbrillig) [reply client/my-sites/checkout/src/components/checkout-main-content.tsx]: I think it was just an unnecessary optimization there too.

**Review by sirbrillig — APPROVED** [review]
Looks good!

Before             |  After
:-------------------------:|:-------------------------:
Screenshot 2024-03-10 at 5 10 48 PM | Screenshot 2024-03-10 at 5 10 56 PM

**Comment by jjchrisdiehl** [comment]
I can confirm that the odd bug in your screenshot above is gone 
image

**Comment by a8ci18n** [comment]
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/11592778

Some locales (Hebrew, Japanese) have been temporarily machine-translated due to translator availability. All other translations are usually ready within a few days. Untranslated and machine-translated strings will be sent for translation next Monday and are expected to be completed by the following Friday.

Thank you @jjchrisdiehl for including a screenshot in the description! This is really helpful for our translators.

**Comment by a8ci18n** [comment]
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/11592778).

**Pull request merged**
